From 4782c7c8466957f89d4f2b8aa1e0f935ebe9f524 Mon Sep 17 00:00:00 2001
From: Michal Schmidt <mschmidt@redhat.com>
Date: Fri, 5 Sep 2014 19:49:24 -0400
Subject: [ethernet] bnx2x: Don't receive packets when the napi budget == 0

Message-id: <1409946615-5862-20-git-send-email-mschmidt@redhat.com>
Patchwork-id: 91561
O-Subject: [RHEL7.1 PATCH 19/70] bnx2x: Don't receive packets when the napi budget == 0
Bugzilla: 1089402
RH-Acked-by: John Linville <linville@redhat.com>
RH-Acked-by: Nikolay Aleksandrov <nikolay@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>

BZ: https://bugzilla.redhat.com/show_bug.cgi?id=1089402

Upstream commit:
commit b3529744b412e9870c9e10fef874e3bee2af1afa
Author: Eric W. Biederman <ebiederm@xmission.com>
Date:   Fri Mar 14 17:57:59 2014 -0700

    bnx2x: Don't receive packets when the napi budget == 0

    Processing any incoming packets with a with a napi budget of 0
    is incorrect driver behavior.

    This matters as netpoll will shortly call drivers with a budget of 0
    to avoid receive packet processing happening in hard irq context.

    Signed-off-by: "Eric W. Biederman" <ebiederm@xmission.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/broadcom/bnx2x/bnx2x_cmn.c b/drivers/net/ethernet/broadcom/bnx2x/bnx2x_cmn.c
index 6a0d17c..702c65b 100644
--- a/drivers/net/ethernet/broadcom/bnx2x/bnx2x_cmn.c
+++ b/drivers/net/ethernet/broadcom/bnx2x/bnx2x_cmn.c
@@ -878,6 +878,8 @@ static int bnx2x_rx_int(struct bnx2x_fastpath *fp, int budget)
  if (unlikely(bp->panic))
   return 0;
 #endif
+ if (budget <= 0)
+  return rx_pkt;
 
  bd_cons = fp->rx_bd_cons;
  bd_prod = fp->rx_bd_prod;
-- 
1.7.1